### DISCUSSION TITLE
Where should schemas live in Python packages?

### DISCUSSION THREAD

**Discussion by t-mart** [python-jsonschema/jsonschema#986] [Q&A] [answered]
I'm writing a CLI tool that will read a configuration file that should conform to a prescribed schema. I've written this schema to a file, and I am trying to figure out:

- Where should it live in my project's directory structure?
- How can I locate it from within code? I'm hoping this would _not_ be a `__file__` relative path because that would depend on the path of the locating module. Maybe there is some semblance of a "package root"?
- Since it is not code, will my package manager (I use Poetry) need to be aware of this file to include it for distribution?
- And finally, will such a solution work in both development (e.g. testing things inside the source code directory) and in installed (e.g. user's `pip install`) use cases?

Thanks!

**Comment by sbrunner** [comment, marked as answer]
Personally, I put it in the code and use `pkgutil.get_data("<module>", "<file>")` to get it.
With Poetry I add:
```
include = [
    "<module>/*.json",
]
```
In the `[tool.poetry]` of the `pyproject.toml`.

I didn't find any issue with dev or production mode with this solution :-)

**Reply by Julian** [reply]
This sounds like decent advice to me too. The only thing I'd add is [`importlib-resources`](https://docs.python.org/3/library/importlib.html#module-importlib.resources) is technically the newer replacement for that `pkgutil` API.

You can see an example of using it for precisely this use case in [jsonschema itself, where it's used to load (meta-)schemas](https://github.com/python-jsonschema/jsonschema/blob/990bae77c368fe885492dede9ef5adfbb950924c/jsonschema/_utils.py#L60-L62) in a private API.
